library/std/src/collections/hash/map.rs

```diff
@@ -2501,6 +2502,58 @@ where
 }
 
 impl<'a, K, V> Entry<'a, K, V> {
+    /// Inserts `value` if the entry is vacant, otherwise merges the existing
+    /// value with `value`, using the given function `f`.
+    ///
+    /// If a panic occurs in the function, the entry will be removed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(hashmap_entry_merge)]
+    ///
+    /// use std::collections::HashMap;
+    /// use std::ops::Add;
+    ///
+    /// let mut word_counts: HashMap<&str, u32> = HashMap::new();
+    /// word_counts.entry("very").merge(1, Add::add);
+    /// word_counts.entry("very").merge(1, Add::add);
+    /// word_counts.entry("wonderful").merge(1, Add::add);
+    ///
+    /// assert_eq!(word_counts["very"], 2);
+    /// assert_eq!(word_counts["wonderful"], 1);
+    /// ```
+    #[unstable(feature = "hashmap_entry_merge", issue = "none")]
+    pub fn merge(self, value: V, f: impl FnOnce(V, V) -> V) -> &'a mut V {
+        struct Guard<'a, K, V> {
+            copied: Option<OccupiedEntry<'a, K, V>>,
+        }
+        impl<K, V> Drop for Guard<'_, K, V> {
+            fn drop(&mut self) {
+                if let Some(entry) = self.copied.take() {
+                    mem::forget(entry.remove());
+                }
+            }
+        }
```

Comment:
I think generally the pattern avoids `Option`s in guards, and instead of `.take()`ing them on success it's generally preferred to `mem::forget` the guard.

Reply:
I don't see a way to do this at least without more unsafe code. The borrow checker won't let me return a reference to something owned by the guard.

Reply:
Oh right, you need ownership here. I'm not sure what's the best way to do this then, you would need extra `unsafe` code, but given that you're already doing `unsafe` stuff...